compiler/native/environment.go

```diff
@@ -17,30 +17,68 @@ import (
 )
 
 // EnvironmentStages injects environment variables
-// for each step in every stage in a yaml configuration.
+// for each stage in a yaml configuration.
 // nolint:lll // ignore function line length
 func (c *client) EnvironmentStages(s yaml.StageSlice, globalEnv raw.StringSliceMap) (yaml.StageSlice, error) {
 	// iterate through all stages
 	for _, stage := range s {
-		// inject the environment variables into the steps for the stage
-		steps, err := c.EnvironmentSteps(stage.Steps, globalEnv)
+		_, err := c.EnvironmentStage(stage, globalEnv)
 		if err != nil {
 			return nil, err
 		}
+	}
 
-		stage.Steps = steps
+	return s, nil
+}
+
+// EnvironmentStage injects environment variables
+// for each stage in a yaml configuration.
+// nolint:lll // ignore function line length
+func (c *client) EnvironmentStage(s *yaml.Stage, globalEnv raw.StringSliceMap) (*yaml.Stage, error) {
+	// make empty map of environment variables
+	env := make(map[string]string)
+	// gather set of default environment variables
+	defaultEnv := environment(c.build, c.metadata, c.repo, c.user)
+
+	// inject the declared global environment
+	// WARNING: local env can override global
+	env = appendMap(env, globalEnv)
+
+	// inject the declared environment
+	// variables to the build step
+	for k, v := range s.Environment {
+		env[k] = v
+	}
+
+	// inject the default environment
+	// variables to the build stage
+	// we do this after injecting the declared environment
+	// to ensure the default env overrides any conflicts
+	for k, v := range defaultEnv {
+		env[k] = v
 	}
```

Comment:
I'm curious as to whether or not this is necessary.  Since the stage will never have a commands block, I don't see a possible scenario where a user overwrites a default environment. However, it might be nice to have this so that if we were to ever introduce some sort of functionality with environment variables, whoever implemented that change would not have to worry about changing this. Would love to hear more thoughts!

Reply:
i'm good with leaving it in as a precautionary measure.